quickwit/quickwit-common/src/semaphore_with_waiter.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use tokio::sync::{OwnedSemaphorePermit, Semaphore, TryAcquireError};
+
+/// `SemaphoreWithLimitedWaiters` is an extension of semaphore
+/// that limits the number of waiters.
+///
+/// If more than n-waiters then acquire returns an error.
+#[derive(Clone)]
+pub struct SemaphoreWithLimitedWaiters {
+    permits: Arc<Semaphore>,
```

Comment:
I would get rid of the two `Arc` and let callers wrap the final semaphore with an `Arc` just like for a regular `Semaphore`.